sql: public schema long running migration

RichardJCai opened the pull request:
sql: public schema long running migration

Release note: None

RichardJCai committed:
sql: public schema long running migration

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/72662)
<!-- Reviewable:end -->

ajwerner commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,218 @@
+	// databases. Is it possible that a user concurrently creates a database
+	// that has a synthetic public schema while this query executes and
+	// that this query doesn't find that database?
+	query := `SELECT "parentID", "parentSchemaID", name, id, crdb_internal_mvcc_timestamp 
+FROM system.namespace WHERE "parentID" != 1 AND "parentSchemaID" = 0 AND name = 'public' ORDER BY "parentID" ASC`
+	rows, err := d.InternalExecutor.QueryIterator(
+		ctx, "get_databases_with_synthetic_public_schemas", nil /* txn */, query,
+	)
```
what about an anti-join and then stream through these instead of buffering? Like:

```sql
  SELECT ns_db.name, ns_db.id, ns_db.crdb_internal_mvcc_timestamp
    FROM system.namespace AS ns_db
         INNER JOIN system.namespace AS ns_sc ON (
                                                                ns_db.id
                                                                = ns_sc."parentID"
                                                             )
   WHERE ns_db.id != 1
     AND ns_db."parentSchemaID" = 0
     AND ns_db."parentID" = 0
     AND ns_sc."parentSchemaID" = 0
     AND ns_sc.name = 'public'
     AND ns_sc.id = 29
ORDER BY ns_db.id ASC;
```

ajwerner replied:
what's the timestamp for?

ajwerner commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,218 @@
+		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
+	) error {
+
+		// Create new schema, go through each table ALTER TABLE SET SCHEMA, rename public schema?
+		row, err := d.InternalExecutor.QueryRow(ctx, "get_database_desc", txn, `
+SELECT id, descriptor, crdb_internal_mvcc_timestamp FROM system.descriptor WHERE id = $1`, dbID)
+		if err != nil {
+			return err
+		}
+		_, desc, _, err := unmarshalDescFromDescriptorRow(row)
+		if err != nil {
+			return err
+		}
+		if err != nil {
+			return err
+		}
+
+		mutableDesc := dbdesc.NewBuilder(desc.GetDatabase())
+		dbDesc := mutableDesc.BuildExistingMutable()
```
use the descs.Collection for this

ajwerner commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,218 @@
+	batch *kv.Batch,
+	newPublicSchemaID descpb.ID,
+) error {
+	// Get All Descriptors under the current database's public schema.
+	rows, err := d.InternalExecutor.QueryIterator(ctx, "get_objects_in_db_public_schema", txn,
+		`SELECT d.id, descriptor, d.crdb_internal_mvcc_timestamp FROM system.descriptor d JOIN system.namespace ns ON d.id = ns.id WHERE ns."parentID" = $1
+AND ns."parentSchemaID" = 29`, dbID)
+	if err != nil {
+		return err
+	}
```
We have some things to retrieve batches of descriptors from IDs and we have some things to get the set of descriptors in a schema. Can we use those instead of this? We don't currently have good batched retrieval in the descs.Collection, and that does make me sad, I'm inclined to ask you to write this using the highest level catalog APIs and then it'll motivate me or somebody to work on batched retrieval.

RichardJCai replied:
Added a function `GetAllDescriptorsInDatabase`. I think this is where we want to add the batching right? I guess we could turn this `GetAllDescriptorsInDatabase` function into an iterator with batch sizes if we're worried about memory.

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+     AND ns_sc."parentSchemaID" = 0
+     AND ns_sc.name = 'public'
+     AND ns_sc.id = 29
+ORDER BY ns_db.id ASC;
```
nit: feed this into sqlfm.pt

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+			publicSchemaDesc,
+		); err != nil {
+			return err
+		}
```
Would it be possible instead to either:
1. create the schema by executing a CREATE SCHEMA statement and then reading and modifying that descriptor subsequently if necessary?
2. create the schema descriptor using `sql.CreateUserDefinedSchemaDescriptor`?

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+			ctx, false, d.Settings, b, d.Codec, dbDesc.GetID(), dbDesc,
+		); err != nil {
+			return err
+		}
```
Can you use the collection's WriteDescToBatch method here? This will call MaybeIncrementVersion for you, among many other things.

RichardJCai replied:
Nice yeah good call

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+			tree.PublicSchema: {
+				ID: publicSchemaID,
+			},
+		}
```
Won't this overwrite any existing Schemas map completely? 

I suggest that in your test, you add a user-defined schema to your database before running the migration, to cover this case.

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+			return err
+		}
+
+		dbDesc.DescriptorProto().GetDatabase().Schemas = map[string]descpb.DatabaseDescriptor_SchemaInfo{
```
nit: because `dbDesc` is mutable, you can actually do `dbDesc.Schemas = ...`. If you'd rather not access the embedded struct's field directly, `dbDesc.DatabaseDescriptor().Schemas = ...` is the way to go.

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
+	) error {
+
+		found, desc, err := descriptors.GetImmutableDatabaseByID(ctx, txn, dbID, tree.DatabaseLookupFlags{})
```
One transaction per database is the way to go, but wouldn't it be safer to also check here also that `desc`'s public schema is indeed descriptorless, and return early if for some reason it now has one? 

Bear in mind that this transaction is not the same as the one that ran your namespace join (and that's fine). I'm not sure if this matters, but even so you're better off checking anyway, because you got this db id from the namespace table, not because you directly know for a fact that this db has no public schema descriptor. It should be equivalent, but who knows if that's always the case. Let's just stay on the safe side and check.

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+		if !found {
+			return errors.Newf("expected to find database with id %d", dbID)
+		}
+		mutableDesc := dbdesc.NewBuilder(desc.DatabaseDesc())
```
nit: `s/mutableDesc/dbDescBuilder/` ?

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+	newPublicSchemaID descpb.ID,
+	descriptors *descs.Collection,
+) error {
+	descs, err := descriptors.GetAllDescriptorsInDatabase(ctx, txn, dbID)
```
Hmm. I looked into this and discussed alternatives with Andrew. We've come to the consensus that what would currently be best would be to call `descriptors.GetAllDescriptors` at the beginning of `createPublicSchemaForDatabase` instead, and use those descriptors for everything.

RichardJCai replied:
I'm a bit unclear about what `use those descriptors for everything` means but I've updated it to call `desriptors.GetAllDescriptors` in `createPublicSchemaForDatabase`  before `migrateObjectsInDatabase` which then iterates through `allDescriptors`. Let me know if this wasn't what you meant.

postamar replied:
Yes this is what I meant, sorry for not being clear.

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+
+	var modifiedDescs []catalog.MutableDescriptor
+	for _, desc := range descs {
+		b := catalogkv.NewBuilder(desc.DescriptorProto())
```
nit: do `b := desc.NewBuilder()` instead. I recently added this method exactly for use cases like these, which have nothing to do with the `catalogkv` package.

RichardJCai replied:
Very nice

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+		case catalog.Database, catalog.Schema:
+			panic(fmt.Sprintf("unexpected descriptor type %v", b.DescriptorType()))
+		case catalog.Table:
+			table := b.BuildExistingMutable()
```
nit: do `switch mut := b.BuildExistingMutable().(type)` instead, and then you can access fields directly, like `mut.UnexposedParentSchemaID = newPublicSchemaID` in `case *tabledesc.Mutable` and `mut.ParentSchemaID = newPublicSchemaID` in `case *typedesc.Mutable`

RichardJCai replied:
Good call, updated

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+			typ.DescriptorProto().GetType().ParentSchemaID = newPublicSchemaID
+			newKey := catalogkeys.MakeObjectNameKey(d.Codec, typ.GetParentID(), typ.GetParentSchemaID(), typ.GetName())
+			batch.Put(newKey, typ.GetID())
+			modifiedDescs = append(modifiedDescs, typ)
```
nit: aside from the `newPublicSchemaID` assignment everything else in these switch-cases can be factored out.

RichardJCai replied:
Done

postamar commented on pkg/migration/migrations/public_schema_migration.go:
```diff
@@ -0,0 +1,203 @@
+	for _, modified := range modifiedDescs {
+		err := catalogkv.WriteDescToBatch(
+			ctx, false, d.Settings, batch, d.Codec, modified.GetID(), modified,
+		)
```
Here also, use `descriptors.WriteDescToBatch` instead if possible, it should be better and safer?

RichardJCai replied:
Done

postamar reviewed (changes requested):
Thanks for doing this! I have a few comments, nothing major I don't think.

postamar commented:
This looks good to me, good work! I'm hoping @ajwerner will also do a review pass, migrations are quite sensitive and are worth being careful about.

RichardJCai commented:
Gonna add one more test using a backup.

ajwerner reviewed (commented):
I'm generally good on this. Just minor things. This is a classic example of something that'd be nice to exercise against real store dirs from the previous version. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-:-MpxUoFc1TQGaSssiFkw:b-iern7q)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @RichardJCai)

___
*[pkg/migration/migrations/public_schema_migration.go, line 72 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-MpxQzXG2WdsnZJrqYE8:-MpxQzXG2WdsnZJrqYE9:b-hasvcj) ([raw file](https://github.com/cockroachdb/cockroach/blob/c05651088e9a1ee5f9e6e94d6da4c21e078174f6/pkg/migration/migrations/public_schema_migration.go#L72)):*
> ```Go
> 
> 	for _, dbID := range databaseIDs {
> 		fmt.Println("dbID:", dbID)
> ```

detritus
___
*[pkg/migration/migrations/public_schema_migration.go, line 86 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-MpxRGEc2qhACDH0n4nY:-MpxRGEc2qhACDH0n4nZ:b-knhqao) ([raw file](https://github.com/cockroachdb/cockroach/blob/c05651088e9a1ee5f9e6e94d6da4c21e078174f6/pkg/migration/migrations/public_schema_migration.go#L86)):*
> ```
> 	return d.CollectionFactory.Txn(ctx, d.InternalExecutor, d.DB, func(
> 		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
> 	) error {
> ```
nit: buy yourself an extra layer of indentation by making the body of this a function all on its own that maybe takes just the `codec, settings, txn, collection, dbID`? Generally good practice to scope down your deps when it's easy. 
___
*[pkg/migration/migrations/public_schema_migration.go, line 129 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-MpxTRCy4UGC4IMR1Cv8:-MpxTRCy4UGC4IMR1Cv9:by9aytd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c05651088e9a1ee5f9e6e94d6da4c21e078174f6/pkg/migration/migrations/public_schema_migration.go#L129)):*
> ```
> 		if desc.DatabaseDesc().Schemas == nil {
> 			dbDesc.Schemas = map[string]descpb.DatabaseDescriptor_SchemaInfo{
> ```
while you're here, thoughts on just adding a method to the `dbdesc.Mutable` which will construct the map if it doesn't exist and then add the entry? Seems bad that we have to do this. 
___
*[pkg/migration/migrations/public_schema_migration.go, line 190 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-MpxU9U8AWfhI4gqYru2:-MpxU9U8AWfhI4gqYru3:b-hhfe1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/c05651088e9a1ee5f9e6e94d6da4c21e078174f6/pkg/migration/migrations/public_schema_migration.go#L190)):*
> ```Go
> 	}
> 	for _, modified := range modifiedDescs {
> 		err := descriptors.WriteDescToBatch(
> ```

I'm not sure how real of a concern this is, but it seems to me like we could end up with a large batch here. It's unlikely to be larger than the maximum size of 16 MiB because, we'll generally folks don't have hundreds of thousands of descriptors in their database, but, it's still a concern that would suck to run into. Thoughts on passing in the `*Txn` and then also having this return a batch, and then, when the batch reaches some size, send it and create a new one. You can test this by lowering the KV max command size to something quite small and then making some descriptors with some biggish strings. This is definitely an edge case concern.  


<!-- Sent from Reviewable.io -->

RichardJCai commented:
We got another small issue I think, restoring a database doesn't actually add an entry for the public schema in the namespace table.

See:
https://github.com/RichardJCai/cockroach/blob/backup_test_missing_public_schema_namespace_entry/pkg/ccl/backupccl/restore_old_version_public_schema_test.go

Notice theres no `52 0 public 29` entry.

Output:
```
0 0 system 1
1 29 descriptor 3
1 29 users 4
1 29 zones 5
1 29 settings 6
1 29 tenants 8
1 29 lease 11
1 29 eventlog 12
1 29 rangelog 13
1 29 ui 14
1 29 jobs 15
1 29 web_sessions 19
1 29 table_statistics 20
1 29 locations 21
1 29 role_members 23
1 29 comments 24
1 29 replication_constraint_stats 25
1 29 replication_critical_localities 26
1 29 replication_stats 27
1 29 reports_meta 28
51 0 public 29
50 0 public 29
1 0 public 29
1 29 namespace 30
1 29 protected_ts_meta 31
1 29 protected_ts_records 32
1 29 role_options 33
1 29 statement_bundle_chunks 34
1 29 statement_diagnostics_requests 35
1 29 statement_diagnostics 36
1 29 scheduled_jobs 37
1 29 sqlliveness 39
1 29 migrations 40
1 29 join_tokens 41
1 29 statement_statistics 42
1 29 transaction_statistics 43
1 29 database_role_settings 44
1 29 tenant_usage 45
1 29 sql_instances 46
1 29 span_configurations 47
0 0 defaultdb 50
0 0 postgres 51
0 0 d 52
52 0 s 53
52 53 t 54
52 29 t 55
```

cc @adityamaru

ajwerner commented:
> We got another small issue I think, restoring a database doesn't actually add an entry for the public schema in the namespace table.

Oof, bet that's going to lead to some odd bugs.

RichardJCai commented:
I think we should make a migration for this as well?

ajwerner commented:
> > > We got another small issue I think, restoring a database doesn't actually add an entry for the public schema in the namespace table.
> > 
> > 
> > Oof, bet that's going to lead to some odd bugs.
> 
> I think we should make a migration for this as well?

Definitely.

RichardJCai commented:
Created an issue for the missing `system.namespace` entry.
https://github.com/cockroachdb/cockroach/issues/73535

RichardJCai committed:
sql: insert missing public schema namespace entry

When restoring a database, a namespace entry for the public
schema was not created.

Release note: None

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72662)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/migration/migrations/public_schema_migration.go, line 136 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-MoFRKWuP2sPpEg7PskJ-r5-136:-Mr41H2n3HAbQrwULeWH:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/0022cbcd75755bcdc561ba9c3f128d4a2816da07/pkg/migration/migrations/public_schema_migration.go#L136)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

nit: because `dbDesc` is mutable, you can actually do `dbDesc.Schemas = ...`. If you'd rather not access the embedded struct's field directly, `dbDesc.DatabaseDescriptor().Schemas = ...` is the way to go.
</blockquote></details>

Done
___
*[pkg/migration/migrations/public_schema_migration.go, line 165 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-MoFRKWuP2sPpEg7PskJ-r5-165:-Mr43P-G2NXmjea7nl3X:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/0022cbcd75755bcdc561ba9c3f128d4a2816da07/pkg/migration/migrations/public_schema_migration.go#L165)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Yes this is what I meant, sorry for not being clear.
</blockquote></details>

Done
___
*[pkg/migration/migrations/public_schema_migration.go, line 72 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-MpxQzXG2WdsnZJrqYE8:-Mr404iiB7MeRO2tM6j9:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/c05651088e9a1ee5f9e6e94d6da4c21e078174f6/pkg/migration/migrations/public_schema_migration.go#L72)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

detritus
</blockquote></details>

Done
___
*[pkg/migration/migrations/public_schema_migration.go, line 86 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-MpxRGEc2qhACDH0n4nY:-Mr412ry1kLdeZ_2Roc1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c05651088e9a1ee5f9e6e94d6da4c21e078174f6/pkg/migration/migrations/public_schema_migration.go#L86)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 	return d.CollectionFactory.Txn(ctx, d.InternalExecutor, d.DB, func(
> 		ctx context.Context, txn *kv.Txn, descriptors *descs.Collection,
> 	) error {
> ```
nit: buy yourself an extra layer of indentation by making the body of this a function all on its own that maybe takes just the `codec, settings, txn, collection, dbID`? Generally good practice to scope down your deps when it's easy. 
</blockquote></details>

Done.
___
*[pkg/migration/migrations/public_schema_migration.go, line 129 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-MpxTRCy4UGC4IMR1Cv8:-Mr43EMd7neQfpcnZH7B:bhqqlbu) ([raw file](https://github.com/cockroachdb/cockroach/blob/c05651088e9a1ee5f9e6e94d6da4c21e078174f6/pkg/migration/migrations/public_schema_migration.go#L129)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> ```
> 		if desc.DatabaseDesc().Schemas == nil {
> 			dbDesc.Schemas = map[string]descpb.DatabaseDescriptor_SchemaInfo{
> ```
while you're here, thoughts on just adding a method to the `dbdesc.Mutable` which will construct the map if it doesn't exist and then add the entry? Seems bad that we have to do this. 
</blockquote></details>

Did this in a separate PR since we have other places where we do this
https://github.com/cockroachdb/cockroach/pull/73942
___
*[pkg/migration/migrations/public_schema_migration.go, line 190 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-MpxU9U8AWfhI4gqYru2:-Mr44ZMA7YYI_TRS0pxb:bbdwya5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c05651088e9a1ee5f9e6e94d6da4c21e078174f6/pkg/migration/migrations/public_schema_migration.go#L190)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

I'm not sure how real of a concern this is, but it seems to me like we could end up with a large batch here. It's unlikely to be larger than the maximum size of 16 MiB because, we'll generally folks don't have hundreds of thousands of descriptors in their database, but, it's still a concern that would suck to run into. Thoughts on passing in the `*Txn` and then also having this return a batch, and then, when the batch reaches some size, send it and create a new one. You can test this by lowering the KV max command size to something quite small and then making some descriptors with some biggish strings. This is definitely an edge case concern.  
</blockquote></details>

Addressed all but this one so far, trying to figure out exactly how to do this.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72662)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/migration/migrations/public_schema_migration.go, line 190 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-MpxU9U8AWfhI4gqYru2:-Mr49OnrEH4Yoj0QuKoh:bl7t9o) ([raw file](https://github.com/cockroachdb/cockroach/blob/c05651088e9a1ee5f9e6e94d6da4c21e078174f6/pkg/migration/migrations/public_schema_migration.go#L190)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

Addressed all but this one so far, trying to figure out exactly how to do this.
</blockquote></details>

Added batching PTAL.


<!-- Sent from Reviewable.io -->

postamar reviewed (changes requested):
Just one minor comment. Looking forward to stamping this. Good work!

> Reviewed 210 of 213 files at r5, 4 of 4 files at r6, 1 of 3 files at r9, 4 of 4 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-:-Mt8fhsQ4fVX2B3YzNfz:b-qhcia)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RichardJCai)

___
*[pkg/migration/migrations/migrations.go, line 62 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-Mt8_4IB8Xz4FNdSDgzV:-Mt8_4IB8Xz4FNdSDgzW:bdckpr1) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9383286808cb786cc9a3da552dd80e9972e495/pkg/migration/migrations/migrations.go#L62)):*
> ```Go
> 	),
> 	migration.NewTenantMigration(
> 		"seed system.span_configurations with configs for existing for existing tenants",
> ```

s/for existing for existing tenants/for existing tenants/
___
*[pkg/migration/migrations/public_schema_migration.go, line 101 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-Mt8_OUS8xOIPJOb1anQ:-Mt8_OUS8xOIPJOb1anR:b-vdpfau) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9383286808cb786cc9a3da552dd80e9972e495/pkg/migration/migrations/public_schema_migration.go#L101)):*
> ```Go
> 	if !found {
> 		return errors.Newf("expected to find database with id %d", dbID)
> 	}
> ```

Piggybacking on Andrew's comment: you don't need to check `found` if you pass the `Required` flag to the `descriptors` API call above. Use `tree.DatabaseLookupFlags{Required: true}` and if it's not found, `err` will be set appropriately, meaning you can ignore `found`. This will also bypass leasing and all that, so it's what you want to do here.
___
*[pkg/migration/migrations/public_schema_migration.go, line 175 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-Mt8cQf4ADgYRy_KCUU0:-Mt8cQf4ADgYRy_KCUU1:bma4g3v) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9383286808cb786cc9a3da552dd80e9972e495/pkg/migration/migrations/public_schema_migration.go#L175)):*
> ```Go
> 	batch := txn.NewBatch()
> 	for _, desc := range allDescriptors {
> 		b := desc.NewBuilder()
> ```

You'll want to move this further down, below the if-continue block which will eliminate most descriptors from further consideration. Under the hood, creating a descriptor builder does a `protoutil.Clone`, which isn't cheap. We've ran into perf issues from this kind of thing before, which is why I'm being careful.


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
Updated, also made a minor change such that we don't run the test that creates 500 tables under race.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-:-Mt91_oP33Wi7D5PIScA:bln236d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[pkg/migration/migrations/migrations.go, line 62 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-Mt8_4IB8Xz4FNdSDgzV:-Mt8tajqDzVWx12jGbB5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9383286808cb786cc9a3da552dd80e9972e495/pkg/migration/migrations/migrations.go#L62)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

s/for existing for existing tenants/for existing tenants/
</blockquote></details>

Done.
___
*[pkg/migration/migrations/public_schema_migration.go, line 101 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-Mt8_OUS8xOIPJOb1anQ:-Mt8tZJh2_KBKTKk0ABC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9383286808cb786cc9a3da552dd80e9972e495/pkg/migration/migrations/public_schema_migration.go#L101)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Piggybacking on Andrew's comment: you don't need to check `found` if you pass the `Required` flag to the `descriptors` API call above. Use `tree.DatabaseLookupFlags{Required: true}` and if it's not found, `err` will be set appropriately, meaning you can ignore `found`. This will also bypass leasing and all that, so it's what you want to do here.
</blockquote></details>

Done.
___
*[pkg/migration/migrations/public_schema_migration.go, line 175 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-Mt8cQf4ADgYRy_KCUU0:-Mt8ttduCn3Um_MWeOBD:bk99mkx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d9383286808cb786cc9a3da552dd80e9972e495/pkg/migration/migrations/public_schema_migration.go#L175)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

You'll want to move this further down, below the if-continue block which will eliminate most descriptors from further consideration. Under the hood, creating a descriptor builder does a `protoutil.Clone`, which isn't cheap. We've ran into perf issues from this kind of thing before, which is why I'm being careful.
</blockquote></details>

Thanks good call 


<!-- Sent from Reviewable.io -->

postamar reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 4 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/72662#-:-Mt9OULB0tB0fQWMuaYw:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner)



<!-- Sent from Reviewable.io -->

RichardJCai commented:
Thanks for the reviews 😃 

bors r=postamar
